cmd/errtrace/toolexec.go

```diff
@@ -0,0 +1,180 @@
+	// In toolexec mode, we're passed the original command + arguments.
+	if len(args) == 0 {
+		return -1, false
+	}
+
+	for _, arg := range args {
+		if arg == "-V=full" {
+			// compile is run first with "-V=full" to get a version number
+			// for caching build IDs.
+			// No TOOLEXEC_IMPORTPATH is set in this case.
+			return cmd.toolExecVersion(args), true
+		}
+	}
+
+	if cmd.Getenv == nil {
+		cmd.Getenv = os.Getenv
+	}
+	// When "-toolexec" is used, the go cmd sets the package being compiled in the env.
+	if pkg := cmd.Getenv("TOOLEXEC_IMPORTPATH"); pkg != "" {
+		return cmd.toolExecRewrite(pkg, args), true
+	}
+
+	return -1, false
+}
+
+func (cmd *mainCmd) toolExecVersion(args []string) int {
+	tool := exec.Command(args[0], args[1:]...)
+	var stdout bytes.Buffer
+	tool.Stdout = &stdout
+	tool.Stderr = cmd.Stderr
+	tool.Env = os.Environ()
+	if err := tool.Run(); err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			return exitErr.ExitCode()
+		}
+
+		fmt.Fprintf(cmd.Stderr, "%v failed: %v", args[0], err)
+		return 1
+	}
+
+	// TODO: This version number should change whenever the rewriting changes.
+	fmt.Fprintf(cmd.Stdout, "%s-errtrace0\n", strings.TrimSpace(stdout.String()))
+	return 0
+}
+
+func (cmd *mainCmd) toolExecRewrite(pkg string, args []string) (exitCode int) {
+	// We only need to modify the arguments for "compile" calls which work with .go files.
+	if !isCompile(args[0]) {
+		return cmd.runOriginal(args)
+	}
+
+	// We only modify files that import errtrace, and have "errtrace.ToolExecInstrument"
+	// However, that requires parsing files to determine, which we want to avoid for stdlib
+	// so use a heuristic to detect stdlib packages -- whether the name contains "."".
+	if !strings.Contains(pkg, ".") {
+		return cmd.runOriginal(args)
+	}
+
+	exitCode, err := cmd.rewriteCompile(pkg, args)
+	if err != nil {
+		cmd.log.Print(err)
+		return 1
+	}
+
+	return exitCode
+}
+
+func (cmd *mainCmd) rewriteCompile(pkg string, args []string) (exitCode int, _ error) {
+	parsed := make(map[string]file)
+	var canRewrite, needRewrite bool
+	for _, arg := range args {
+		if !isGoFile(arg) {
+			continue
+		}
+
+		contents, err := os.ReadFile(arg)
+		if err != nil {
+			return -1, errtrace.Wrap(err)
+		}
+
+		f, err := cmd.parseFile(arg, contents)
+		if err != nil {
+			return -1, errtrace.Wrap(err)
+		}
+		parsed[arg] = f
+
+		// TODO: Support an "unsafe" mode to rewrite packages without errtrace imports.
+		if f.importsErrtrace {
+			canRewrite = true
+		}
+		if len(f.inserts) > 0 {
+			needRewrite = true
+		}
+	}
+
+	if !canRewrite || !needRewrite {
+		return cmd.runOriginal(args), nil
+	}
+
+	// Use a temporary directory per-package that is rewritten.
+	tempDir, err := os.MkdirTemp("", filepath.Base(pkg))
+	if err != nil {
+		return -1, errtrace.Wrap(err)
+	}
+
+	newArgs := make([]string, 0, len(args))
+	for _, arg := range args {
+		if !isGoFile(arg) {
+			newArgs = append(newArgs, arg)
+			continue
+		}
+
+		f, ok := parsed[arg]
```

Comment:
nit:
I don't like repeating the logic of what files are in scope here.
Recalculating the same information here smells a little.
The code block above concretely produces (or at least it should) all the information we need: an exact list of arguments to replace. We should use just that, and not recompute.

If relying on a `map[arg]file` like that feels iffy to you, we can replace it with a `[]*file` of `len == len(args)` with `nil` for cases where we don't replace the argument.

Reply:
Good call, I meant to remove the `isGoFile` here and rely on the map, updated.